File: 2713/main.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(maxIncreasingCells([][]int{
		{3,1},
		{3,4},
	}))
}

func maxIncreasingCells(mat [][]int) int {
	m, n := len(mat), len(mat[0])
	mp := make(map[int][][2]int)
	row := make([]int, m)
	col := make([]int, n)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			mp[mat[i][j]] = append(mp[mat[i][j]], [2]int{i, j})
		}
	}
	keys := []int{}
	for key, _ := range mp {
		keys = append(keys, key)
	}
	sort.Ints(keys)

	for _, key := range keys {
		pos := mp[key]
		res := []int{}
		for _, arr := range pos {
			res = append(res, max(row[arr[0]], col[arr[1]]) + 1)
		}
		for i := 0; i < len(pos); i++ {
			arr, d := pos[i], res[i]
			row[arr[0]] = max(row[arr[0]], d)
			col[arr[1]] = max(col[arr[1]], d)
		}
	}

	return maxSlice(row)
}

func maxSlice(slice []int) int {
	maxVal := slice[0]
	for _, val := range slice {
		if val > maxVal {
			maxVal = val
		}
	}
	return maxVal
}

File: 2713/main.rs
use std::collections::HashMap;

struct Solution{}

impl Solution {
    pub fn max_increasing_cells(mat: Vec<Vec<i32>>) -> i32 {
        let m = mat.len();
        let n = mat[0].len();
        let mut mp: HashMap<i32, Vec<(usize, usize)>> = HashMap::new();
        let mut row = vec![0; m];
        let mut col = vec![0; n];

        for i in 0..m {
            for j in 0..n {
                mp.entry(mat[i][j]).or_insert(Vec::new()).push((i, j));
            }
        }

        let mut sorted_map: Vec<_> = mp.iter().collect();
        sorted_map.sort_by(|a, b| a.0.cmp(b.0));
        for (_, pos) in sorted_map {
            let res: Vec<_> = pos.iter().map(|&(i, j) | row[i].max(col[j])+1).collect();
            for (&(i, j), &d) in pos.iter().zip(res.iter()) {
                row[i] = row[i].max(d);
                col[j] = col[j].max(d);
            }
        }
        *row.iter().max().unwrap()
    }
}

fn main() {
    println!("{}", Solution::max_increasing_cells(
        vec![
            vec![3, 1],
            vec![3, 4]
        ]
    ))
}